src/act.movement.cpp

```diff
@@ -679,7 +679,8 @@ void move_vehicle(struct char_data *ch, int dir)
     send_to_char("You aren't the Kool-Aid Man, so you decide against ramming your way out of here.\r\n", ch);
     return;
   }
-  if (!EXIT(veh, dir) || !EXIT(veh, dir)->to_room || EXIT(veh, dir)->to_room == &world[0] || IS_SET(EXIT(veh, dir)->exit_info, EX_CLOSED)) {
+  if (!EXIT(veh, dir) || !EXIT(veh, dir)->to_room || EXIT(veh, dir)->to_room == &world[0] || (IS_SET(EXIT(veh, dir)->exit_info, EX_CLOSED) &&
+  !(ROOM_FLAGGED(EXIT(veh, dir)->to_room, ROOM_HOUSE) && House_can_enter(ch, EXIT(veh, dir)->to_room->number) && has_key(ch, (EXIT(veh, dir)->key))))) {
```

Comment:
You should probably move this added logic to a boolean variable and use that variable here: `bool can_enter_closed_apartment = (your logic)`. Then you can also use that same variable in an if-check to print a message to both the vehicle and the room about opening / closing the door as it passes through.